components/omega/src/infra/Dimension.h

```diff
@@ -0,0 +1,158 @@
+   // Checks to see if a dim of this name exists
+   static bool
+   exists(const std::string &Name ///< [in] name of dimension to check
+   );
+
+   //---------------------------------------------------------------------------
+   // Destroys a dimension
+   static void
+   destroy(const std::string Name ///< [in] name of dimension to destroy
+   );
+
+   //---------------------------------------------------------------------------
+   // Destroys all defined dimensions
+   static void clear();
+
+   //---------------------------------------------------------------------------
+   // Retrieves full dimension instance by name
+   static std::shared_ptr<Dimension>
+   get(const std::string Name // [in] Name of dimension
+   );
+
+   //---------------------------------------------------------------------------
+   /// Get name of dimension from instance (iterator)
+   std::string getName();
+
+   //---------------------------------------------------------------------------
+   /// Check distributed property of a dimension from instance
+   bool isDistributed();
+
+   //---------------------------------------------------------------------------
+   /// Check distributed property of dimension given its name
+   static bool
+   isDistributedDim(const std::string &Name ///< [in] name of dimension
+   );
+
+   //---------------------------------------------------------------------------
+   /// Get global dimension from instance
+   I4 getLengthGlobal();
+
+   //---------------------------------------------------------------------------
+   /// Get global dimension length by name
+   static I4
+   getDimLengthGlobal(const std::string &Name ///< [in] name of dimension
+   );
+
+   //---------------------------------------------------------------------------
+   /// Get length of dimension in local partition from instance
+   I4 getLengthLocal();
+
+   //---------------------------------------------------------------------------
+   /// Get length of dimension in local partition by name
+   static I4
+   getDimLengthLocal(const std::string &Name ///< [in] name of dimension
+   );
+
+   //---------------------------------------------------------------------------
+   /// Get global offset for each local address from dim instance
+   HostArray1DI4 getOffset();
+
+   //---------------------------------------------------------------------------
+   /// Get global offset for each local address given a dim name
+   static HostArray1DI4
+   getDimOffset(const std::string &Name ///< [in] name of dimension
+   );
+
+   //----------------------------------------------------------------------------//
+   // Retrieves the total number of currently defined dimensions
+   static int getNumDefinedDims();
+
+   //---------------------------------------------------------------------------
+   /// An iterator can be used to loop through all defined dimensions
+   using Iter =
+       std::map<std::string, std::shared_ptr<Dimension>>::const_iterator;
```

Comment:
The const_iterator was just adding further enforcement of the read-only aspect of a dimension in this context, but I'll go ahead and make the change since, as you say, it's already enforced by the API.